keps/sig-node/3317-user-namespaces/README.md

```diff
@@ -0,0 +1,1015 @@
+proposal, doesn't have all the feedback and adjustments we discussed) but can
+still be useful as a general overview.
+
+TODO: move the table to markdown and include it here.
+
+
+#### Phase 1: pods "without" volumes
+
+This phase makes pods "without" volumes work with user namespaces. This is
+activated via the bool `pod.spec.HostUsers` and can only be set to `false`
+on pods which use either no volumes or only volumes of the following types:
+
+ - configmap
+ - secret
+ - downwardAPI
+ - emptyDir
+ - projected
+
+This list of volumes was chosen as they can't be used to share files with other
+pods.
+
+The mapping length will be 65535, mapping the range 0-65534 to the pod. This wide
+range makes sure most workloads will work fine.
+
+The mapping will be chosen by the kubelet, using a simple algorithm to give
+different pods in this category ("without" volumes) a non-overlapping mapping.
+Giving non-overlapping mappings generates the best isolation for pods.
+
+Furthermore, the node UID space of 2^32 can hold up to 2^16 pods each with a
+mapping length of 65535 (2^16-1) top. This imposes a limit of 65k pods per node,
+but that is not an issue we will hit in practice for a long time, if ever (today
+we run 110 pods per node by default).
+
+During phase 1, to make sure we don't exhaust the host UID namespace, we will
+limit the number of pods using user namespaces to `min(maxPods, 1024)`. This
+leaves us plenty of host UID space free and this limits is probably never hit in
+practice. See UNRESOLVED for more some UNRESOLVED info we still have on this.
+
+##### pkg/volume changes for phase I
+
+We need the files created by the kubelet for these volume types to be readable
+by the user the pod is mapped to in the host. Luckily, Kuberentes already
+supports specifying an [fsUser and fsGroup for volumes][volume-mounter-args] so
+the changes needed here are rather small.
+
+First, we need to teach the [operation executor][operation-executor] to convert
+the container UIDs/GIDs to the host UIDs/GIDs that this pod is mapped to. To do
+this, we will modify the [volumeHost interface][volumeHost-interface] adding
+functions to transform a container ID to the corresponding host ID, so the
+operation executor will just call that function (via
+self.volumePluginMgr.Host.GetHostIDsForPod())
+
+This function will call the kubelet, which will use the user namespace manager
+created to transform the IDs for that pod. The [volumeHost
+interface][volumeHost-interface] will then have this new method:
+
+```
+GetHostIDsForPod(pod *v1.Pod, containerUID, containerGID *int64) (hostUID, hostGID *int64, err error)
+```
+
+This method will only transform the IDs if the pod is running with userns
+enabled. If userns is disabled, the params will not be translated and, also, 
+GetHostIDsForPod() will return nil if the containerUID/containerGID received
+where nil. This is to keep the current functionality untouched, as [fsUser and
```

Comment:
typo
```suggestion
were nil. This is to keep the current functionality untouched, as [fsUser and
```

Reply:
Fixed locally. Will not push to not lose the LGTM/APPROVE labels, but will be in the next push :)

Thanks!